Build for MacOS

eltorio opened the pull request:
I'm now here!
Ronan

eltorio committed:
Build for MacOS

eltorio commented:
Dear Sylvain,
I have now a working Metal implementation, I have some questions about your CUDA implementation
````c++
     float fov = p_Fov[i]; // Why an array with index 0<=i<samples?
//………
     const float3 r012 = {r[i*9+0], r[i*9+1], r[i*9+2]}; // idem here rotation matrix is indexed on samples?
     const float3 r345 = {r[i*9+3], r[i*9+4], r[i*9+5]};
     const float3 r678 = {r[i*9+6], r[i*9+7], r[i*9+8]};
````
Ronan

eltorio committed:
Working Apple Metal version

eltorio commented:
In depth there is a for loop (i=0;i<samples;i++)
and computation are made with p_Fov[i] and p_RotMat[i*9+0]…[I*9+9] for me it does not make sense since the only multiple fov and rotation matrix I see are with the different cameras (1->20) ,
If you use samples=1 no pb i=0 so it is the p_Fov and the rotation matrix but if you use 2 samples ? what is *(p_Fov+1) I did not see any memory allocation for that, only:
private:
float* _fov;

I’m probably wrong but I need to understand
Ronan

> Le 20 avr. 2021 à 17:20, Sylvain Gravel ***@***.***> a écrit :
> 
> 
> I didn't check the CUDA implementation in depth, I only fixed the black pixel in the center. The implementation comes from Stefan, I'll have to look into the details.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/LRP-sgravel/reframe360XL/pull/1#issuecomment-823361224>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABVE3IJ2FNJKZHFTLBCMVHTTJWLUFANCNFSM43FX54ZA>.
>

eltorio committed:
Support for MacOS Universal app (Apple Silicon ARM64/Intel x86_64)

eltorio committed:
Binary link for MacOS

eltorio committed:
Motion blur support with Apple Metal

eltorio committed:
update MacOS binary link

eltorio committed:
Build OpenCL kernel from source code

Commit:
Merge branch 'master' into master

LRP-sgravel commented:
Merci pour ton travail!!

LRP-sgravel commented:
> 
> 
> Dear Sylvain,
> I have now a working Metal implementation, I have some questions about your CUDA implementation
> 
> ```c++
>      float fov = p_Fov[i]; // Why an array with index 0<=i<samples?
> //………
>      const float3 r012 = {r[i*9+0], r[i*9+1], r[i*9+2]}; // idem here rotation matrix is indexed on samples?
>      const float3 r345 = {r[i*9+3], r[i*9+4], r[i*9+5]};
>      const float3 r678 = {r[i*9+6], r[i*9+7], r[i*9+8]};
> ```
> 
> Ronan

So the samples are used for motion blur.  Those arrays are properly assigned and freed in `RunCudaKernel` in `CudaKernel.cu`.  Motion blur takes n samples around the current frame based on the animation parameters (FOV and transform matrix) and blends them.  The code looks alright but I'm not sure I'll keep it as I would expect motion blur to sample many frames whereas the current implementation uses n samples offset around [-1, 1] frames.  I.E. For 5 samples around frame f, it would blend frames at timestamp f-0.666, f-0.333, f, f+0.333, f+0.666 yielding very limited actual motion blur.

eltorio commented:
Metal version is very fast comparing to OpenCL,

I understand now the use of the index, I modified the Metal code for using it,
I still do not understand how to enable it,
I put a small debug code for printing out the values sent to the compute kernel and If I put 4 samples, I have 4 times the same values
So looping 4 times on the same values is just time consuming so obviously is not the good way to enable it !!!!!

I look deeper in the code later for yet I have to finish another thing
OpenCLKernel Working for, W:3840 H:2160 samples:4 bilinear:1 
	Run 0
fov:1.000000 tinyplanet:0.000000 rectilinear:0.000000 
	rotMat[0]
	         :0.7459     0.1524     -0.6483     
	         :0.0000     0.9735     0.2289     
	         :0.6660     -0.1707     0.7261     
	Run 1
fov:1.000000 tinyplanet:0.000000 rectilinear:0.000000 
	rotMat[1]
	         :0.7459     0.1524     -0.6483     
	         :0.0000     0.9735     0.2289     
	         :0.6660     -0.1707     0.7261     
	Run 2
fov:1.000000 tinyplanet:0.000000 rectilinear:0.000000 
	rotMat[2]
	         :0.7459     0.1524     -0.6483     
	         :0.0000     0.9735     0.2289     
	         :0.6660     -0.1707     0.7261     
	Run 3
fov:1.000000 tinyplanet:0.000000 rectilinear:0.000000 
	rotMat[3]
	         :0.7459     0.1524     -0.6483     
	         :0.0000     0.9735     0.2289     
	         :0.6660     -0.1707     0.7261     


I note strange thing OpenCLKernel.h is not up to date with OpenCLKernel.cl, I modified the Makefile for regenerating from .cl . 

I’m very happy I found a way to use my GoPro Max !

> Le 22 avr. 2021 à 13:29, Sylvain Gravel ***@***.***> a écrit :
> 
> 
> Dear Sylvain,
> I have now a working Metal implementation, I have some questions about your CUDA implementation
> 
>      float fov = p_Fov[i]; // Why an array with index 0<=i<samples?
> //………
>      const float3 r012 = {r[i*9+0], r[i*9+1], r[i*9+2]}; // idem here rotation matrix is indexed on samples?
>      const float3 r345 = {r[i*9+3], r[i*9+4], r[i*9+5]};
>      const float3 r678 = {r[i*9+6], r[i*9+7], r[i*9+8]};
> Ronan
> 
> So the samples are used for motion blur. Those arrays are properly assigned and freed in RunCudaKernel in CudaKernel.cu. Motion blur takes n samples around the current frame based on the animation parameters (FOV and transform matrix) and blends them. The code looks alright but I'm not sure I'll keep it as I would expect motion blur to sample many frames whereas the current implementation uses n samples offset around [-1, 1] frames. I.E. For 5 samples around frame f, it would blend frames at timestamp f-0.666, f-0.333, f, f+0.333, f+0.666 yielding very limited actual motion blur.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/LRP-sgravel/reframe360XL/pull/1#issuecomment-824758797>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABVE3IPWM6XF3FFOQB4SH63TKACCDANCNFSM43FX54ZA>.
>

LRP-sgravel commented:
You need to put the Shutter Angle to something greater than 0 for motion blur to have an effect.  Shutter Angle determines how long the motion blur « trail » will be.  1 = 180 degrees, I will probably change that value to degrees later.
